### PR TITLE
Add PkEncryption support

### PR DESCRIPTION
**Pull request by MatMaul**
Currently a draft since it relies on #8 and a new vodozemac release.

### CODE REVIEW

File: src/pk_encryption.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        let key: &[u8; 32] = bytes
+            .try_into()
+            .map_err(|_| PkEncryptionError::InvalidKeySize(bytes.len()))?;
+
+        Ok(Self {
+            inner: vodozemac::pk_encryption::PkDecryption::from_bytes(&key),
+        })
+    }
+
+    // The secret key used to decrypt messages.
```
Comment (poljar) [inline src/pk_encryption.rs:45]: That ain't a doc comment.

File: src/pk_encryption.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    #[getter]
+    pub fn key(&self) -> Py<PyBytes> {
+        Python::with_gil(|py| PyBytes::new_bound(py, self.inner.to_bytes().as_slice()).into())
+    }
+
+    // The public key used to encrypt messages for this decryption object.
```
Comment (poljar) [inline src/pk_encryption.rs:51]: Neither is this a doc comment.

File: src/pk_encryption.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
+use pyo3::{
+    pyclass, pymethods,
+    types::{PyBytes, PyType},
+    Bound, Py, Python,
+};
+use vodozemac::Curve25519PublicKey;
+
+use crate::PkEncryptionError;
+
+#[pyclass]
```
Comment (poljar) [inline src/pk_encryption.rs:10]: Docs for this, PyO3 transforms the Rust docs into Python ones, I know I know, the rest of the bindings are missing docs as well, but that's something we need to work on as well.

File: Cargo.toml [hunk 4]
```diff
@@ -7,10 +7,14 @@ edition = "2021"
 name = "vodozemac"
 crate-type = ["cdylib"]
 
+[features]
```
Comment (poljar) [inline Cargo.toml:10]: I don't think we need to re-expose the feature flag. Just enable the feature.

**Review by poljar — CHANGES_REQUESTED** [review]

File: src/pk_encryption.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
+//! ☣️  Compat support for libolm's PkEncryption and PkDecryption
```
Comment (MatMaul) [inline src/pk_encryption.rs:1]: I am not sure this will get generated properly in the doc, and I am not sure how I can generate the doc to check.

Comment (MatMaul) [reply src/pk_encryption.rs:1]: And `pk_encryption` is not registered a Python module. We could do so but unsure about the benefits.

I am thinking about moving this warning to the docstrings of `PkDecryption` and `PkEncryption`.

**Comment by poljar** [comment]
Let's undraft this, we can depend on a git rev of vodozemac for a bit before I do the release.

**Comment by MatMaul** [comment]
As a first I've undraft #8 and pin the last git revision of vodozemac.

**Comment by poljar** [comment]
Can this be undrafted now as well?

**Comment by MatMaul** [comment]
I rebased on main and undrafted.

File: src/types/curve25519.rs [hunk 6]
```diff
@@ -22,14 +26,101 @@ impl Curve25519PublicKey {
...
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        Curve25519PublicKey {
+            inner: vodozemac::Curve25519PublicKey::from(&self.inner),
+        }
+    }
+
+    #[classattr]
+    const __hash__: Option<PyObject> = None;
+
+    fn __eq__(&self, other: &Self) -> bool {
```
Comment (poljar) [inline src/types/curve25519.rs:123]: Is this needed for something? This potentially introduces a timing attack.

File: src/error.rs [hunk 7]
```diff
@@ -139,3 +139,29 @@ impl From<PickleError> for PyErr {
         PickleException::new_err(e.to_string())
     }
 }
+
+#[derive(Debug, Error)]
```
Comment (poljar) [inline src/error.rs:143]: Docs please.

File: src/types/curve25519.rs [hunk 8]
```diff
@@ -22,14 +26,101 @@ impl Curve25519PublicKey {
...
     const __hash__: Option<PyObject> = None;
 
     fn __eq__(&self, other: &Self) -> bool {
         self.inner == other.inner
     }
 }
+
+#[pyclass]
+#[derive(Clone)]
+pub struct Curve25519SecretKey {
```
Comment (poljar) [inline src/types/curve25519.rs:62]: This class and methods also need docs.

**Review by poljar — CHANGES_REQUESTED** [review]
Left a couple of nits, and clippy needs to be convinced as well.

File: src/types/curve25519.rs (hunk 6 above)
Comment (MatMaul) [reply src/types/curve25519.rs:123]: I don't really know, I copy-pasted `Curve25519PublicKey` to be honest :innocent: 

It seems useless for both ? Should I also remove it from `Curve25519PublicKey` ?

Comment (poljar) [reply src/types/curve25519.rs:123]: Comparing public keys is fine, you even might want to have a hashmap or set where the public keys act as hash keys.

For secret keys this isn't true.

Let's just remove it for the secret key.

**Pull request merged**